src/rules_pygen/rules_generator.py

```diff
@@ -0,0 +1,499 @@
+            )
+            for additional_filename, additional_link in wheel_links.items():
+                if additional_filename.startswith(match_prefix):
+                    # we've found a wheel with the same name+version, obviously
+                    # it could just be the wheel we downloaded (the local wheel)
+                    # so we check for that first.
+                    logger.info("Found additional wheel: %s", additional_filename)
+
+                    is_compatible = _check_compatibility(additional_filename)
+                    if not is_compatible:
+                        continue
+
+                    filepath = os.path.abspath(os.path.join(self.wheel_dir, additional_filename))
+                    logger.debug('Considering %s', additional_filename)
+
+                    if wheel_filename != additional_filename:
+                        logger.debug('%s does not equal %s', wheel_filename, additional_filename)
+                        _download(additional_link, filepath)
+
+                    logger.debug('Matched %s %s', match_prefix, additional_filename)
+
+                    wi = WheelInfo(
+                        name=name, filepath=filepath, url=additional_link, version=version,
+                    )
+
+                    dependency.add_wheel(wi)
+
+            if dependency.name not in BLACKLIST:
+                if not dependency.verify({'macos', 'linux'}):
+                    raise PyBazelRuleGeneratorException("Dependency {} is missing wheels!".format(dependency))
+                all_deps.add(dependency)
+        return all_deps
+
+    def _gen_output_file(self, deps) -> None:
+        """Output a file with the following structure
+
+        def pypi_libraries():
+            py_library(
+                name="virtualenv",
+                deps=[
+                ] + ["@pypi_virtualenv//:pkg"],
+                licenses=["notice"],
+                visibility=["//visibility:public"],
+            )
+
+        def pypi_archives():
+            existing_rules = native.existing_rules()
+            if "pypi_asn1crypto" not in existing_rules:
+                http_archive(
+                    name="pypi_asn1crypto",
+                    url="https://files.pythonhosted...",
+                    sha256="2f1adbb7546ed199e3c90ef23ec95c5cf3585bac7d11fb7eb562a3fe89c64e87",
+                    build_file_content=_BUILD_FILE_CONTENT,
+                    type="zip",
+                )
+        """
+        f = tempfile.NamedTemporaryFile(delete=False, mode='w+t')
+        # sort the deps for better diffs
+        sorted_deps = list(deps)
+        sorted_deps.sort(key=operator.attrgetter('name'))
+
+        # header
+        f.write(HEADER)
+        f.write('\ndef pypi_libraries():\n\n')
+
+        # py_libraries
+        for dependency in sorted_deps:
+            logger.debug('Writing py_library for %s', dependency)
+            f.write('    native.py_library(\n')
```

Comment:
Instead of typing spaces explicitly, you could have a `def _space(x): return ' ' * x` helper, and then do stuff like
```
f.write(_space(4) + 'native.py_library(\n')
```
I think it would make things cleaner, especially below where you start needing more and more spaces

Reply:
nice idea, thx